pkg/perf/publicip/publicip_task.go

```diff
@@ -0,0 +1,322 @@
+	}
+}
+
+func (p *publicIPValidationTask) ID() string {
+	return p.taskID
+}
+
+func (p *publicIPValidationTask) Cron() string {
+	return p.schedule
+}
+
+func (p *publicIPValidationTask) Run(ctx context.Context) (interface{}, error) {
+
+	netNS, err := namespace.GetByName(testNamespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get namespace %s: %w", testNamespace, err)
+	}
+	manager, err := environment.GetSubstrate()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get substrate client: %w", err)
+	}
+	sub, err := manager.Substrate()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get substrate client: %w", err)
+	}
+	defer sub.Close()
+	farmID := environment.MustGet().FarmID
+
+	shouldRun, err := isLeastValidNode(ctx, uint32(farmID), sub)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if the node should run public IP verification: %w", err)
+	}
+	if !shouldRun {
+		log.Info().Msg("skipping because there is a node with less ID available")
+		return nil, nil
+	}
+
+	farm, err := sub.GetFarm(uint32(farmID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get farm with id %d: %w", farmID, err)
+	}
+	deleteOldIPs(farm.PublicIPs, p.farmIPsReport)
+	err = netNS.Do(func(_ ns.NetNS) error {
+		return p.validateIPs(farm.PublicIPs)
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to run public IP validation: %w", err)
+	}
+	return p.farmIPsReport, nil
+}
+
+func (p *publicIPValidationTask) validateIPs(publicIPs []substrate.PublicIP) error {
+	mv, err := macvlan.GetByName(testMacvlan)
+	if err != nil {
+		return fmt.Errorf("failed to get macvlan %s in namespace %s: %w", testMacvlan, testNamespace, err)
+	}
+	// to delete any leftover IPs or routes
+	err = deleteAllIPsAndRoutes(mv)
+	if err != nil {
+		log.Err(err).Send()
+	}
+
+	for _, publicIP := range publicIPs {
+		if report, ok := p.farmIPsReport[publicIP.IP]; ok && report.State == ValidState {
+			// no need to test it again
+			continue
+		}
+		p.farmIPsReport[publicIP.IP] = IPReport{
+			State: ValidState,
+		}
+		if publicIP.ContractID != 0 {
+			p.farmIPsReport[publicIP.IP] = IPReport{
+				State:  SkippedState,
+				Reason: IPIsUsed,
+			}
+			continue
+		}
+
+		ip, ipNet, routes, err := getIPWithRoute(publicIP)
+		if err != nil {
+			p.farmIPsReport[publicIP.IP] = IPReport{
+				State:  InvalidState,
+				Reason: PublicIPDataInvalid,
+			}
+			log.Err(err).Send()
+			continue
+		}
+		err = macvlan.Install(mv, nil, ipNet, routes, nil)
+		if err != nil {
+			p.farmIPsReport[publicIP.IP] = IPReport{
+				State:  InvalidState,
+				Reason: PublicIPDataInvalid,
+			}
+			log.Err(err).Msgf("failed to install macvlan %s with ip %s to namespace %s", testMacvlan, ipNet, testNamespace)
+			continue
+		}
+
+		realIP, err := getRealPublicIP()
+		if err != nil {
+			p.farmIPsReport[publicIP.IP] = IPReport{
+				State:  SkippedState,
+				Reason: FetchRealIPFailed,
+			}
+			log.Err(err).Msg("failed to get node real IP")
+			continue
+		}
```

Comment:
This is not always correct, if you failed to get the real ip this can duo to different reasons. and they all need to be handled differently:
- http.Get fail, this can be duo to name look failure, or failed to establish connection this is hugely mean that the IP is invalid (is not routable to the internet) 
- http.Get succeeded but the response is not with success status (!= 200) this means that u have internet but the check service is donw, in that case we are not sure about the IP, we can mark it as skipped, for now.
- http.Get succeed but the the returned value does not match the actual IP we are testing means those IPs are natted, so they are also invalid.

the best way to do this is that getRealPublicIP need to return (wrapped) errors with more details on what happened exactly, and then based on that choose the proper state